mantidproject/mantid
# Accomodate M4 for LOQ in ISIS SANS reduction

Fixes #15664

**Context**

The LOQ instrument at ISIS will be incorporating an M4 monitor at `S17788`. We can load these data sets (with the right IDF) into Mantid now. In addition we need to be able to process them in the ISIS SANS reduction setup. 

This ticket needs to be signed off by the instrument scientists before we can test it. Also it should only be added when the M4 monitors are being used routinely.
## For testing:

There are two test folders located here: smb://olympic/babylon5/Scratch/Anton/Testing

**Reducing non-M4**

Include the folder `LOQNonM4 to the Mantid path.
- Open the ISIS SANS GUI.
- Set the instrument to LOQ
- Load the following data:

| Type | Sans | Transmission | Direct |
| --- | --- | --- | --- |
| Sample: | 93387 | 93385 | 93383 |
| Can: | 93386 | 93384 | 93383 |
- Press `Load`
  - Open the meta information in MantidDock for `93387_sans_nxs`
    - Confirm that _Instrument from_ is: `LOQ_Definition_20121016-.xml`
    - Confirm that _Parameters from_ is: `LOQ_Parameters.xml`
  - Open the meta information in MantidDock for `93386_sans_nxs`
    - Confirm that _Instrument from_ is: `LOQ_Definition_20121016-.xml`
    - Confirm that _Parameters from_ is: `LOQ_Parameters.xml`  
  - Open the meta information in MantidDock for `93385_trans_nxs`
    - Confirm that _Instrument from_ is: `LOQ_trans_Definition.xml`
  - Open the meta information in MantidDock for `93384_trans_nxs`
    - Confirm that _Instrument from_ is: `LOQ_trans_Definition.xml`
  - Open the meta information in MantidDock for `93383_trans_nxs`
    - Confirm that _Instrument from_ is: `LOQ_trans_Definition.xml`
- Press `1D Reduce`
  - Confirm that there are no errors

**Reducing M4**

Include the folder `LOQNonM4 to the Mantid path.
- Open the ISIS SANS GUI.
- Set the instrument to LOQ
- Load the following data:

| Type | Sans | Transmission | Direct |
| --- | --- | --- | --- |
| Sample: | 95593 | 95593 | 95594 |
- Press `Load`
  - Open the meta information in MantidDock for `95593_sans_nxs`
    - Confirm that _Instrument from_ is: `LOQ_Definition_20151016-.xml`
    - Confirm that _Parameters from_ is: `LOQ_Parameters.xml`
  - Open the meta information in MantidDock for `95593_trans_nxs`
    - Confirm that _Instrument from_ is: `LOQ_trans_Definition_M4.xml`
  - Open the meta information in MantidDock for `95594_trans_nxs`
    - Confirm that _Instrument from_ is: `LOQ_trans_Definition_M4.xml`
- Press `1D Reduce`
  - Confirm that there are no errors
- Go to ReductionSettings > Transmission Settings and select M4
- Press `1D Reduce`
  - Confirm that there are no errors
## Release notes

Check in `Files Changed` that the `sans.rst` has an addition
# Additional issues and test instructions
## Empty Bin Issue

The zero-value bins at the end of the unfitted data set which is obtained from the `CalculateTransmission` algorithm are actually correct. A combination of flatbackground correction and rebinning causes the last two bins to be 0 for both the transmission and the direct workspaces (for the 17788 spectrum). When they are divided by each other, this results in a _nan_ which is set to 0 by `ReplaceSpecialValues`. Stephen and I looked at how SANS2D does it and it only performs a flat background subtraction of the M2 monitor. Performing only the flat background subtraction on M2 resolve the issue of the zero-valued bins.
### Test for zero-valued bins
- Use the user file  [USER_LOQ_163A_M4Test_Changer_with_M2_flat_background.txt](https://github.com/mantidproject/mantid/files/482974/USER_LOQ_163A_M4Test_Changer_with_M2_flat_background.txt)
  to perform the reduction with:

| Type | Sans | Transmission | Direct |
| --- | --- | --- | --- |
| Sample: | 98128 | 98128 | 98131 |
- Confirm that the last two bins of 98128_trans_sample_2.2_10.0_unfitted  are not 0

Note that the only change in the user file is:

```
!BACK/MON/TIMES 31000 39000
BACK/M2 31000 39000
```
## M4 checkbox issue
### Test that M4 reduction can be selected after M3-based user file has been loaded
- Load the user file [USER_LOQ_163A_M3_Changer.txt](https://github.com/mantidproject/mantid/files/482986/USER_LOQ_163A_M3_Changer.txt)
- Set the following files:

| Type | Sans | Transmission | Direct |
| --- | --- | --- | --- |
| Sample: | 98128 | 98128 | 98131 |
- Load the files
- Go to the `Reduction Settings` tab and confirm that `Transmission Settings`>M3 is selected
-  Check the box `Transmission Settings`>M4
- Go back to the `Run Numbers` tab and click `Reduce 1D`
  - Confirm that the reduction completes without any issues
### Test that M4 reduction is correctly set when M4-based user file is loaded
- Load the user file  [USER_LOQ_163A_M4Test_Changer_with_M2_flat_background.txt](https://github.com/mantidproject/mantid/files/482974/USER_LOQ_163A_M4Test_Changer_with_M2_flat_background.txt)
- Set the following files:

| Type | Sans | Transmission | Direct |
| --- | --- | --- | --- |
| Sample: | 98128 | 98128 | 98131 |
- Load the files
- Go to the `Reduction Settings` tab and confirm that `Transmission Settings`>M4 is selected
- Go back to the `Run Numbers` tab and click `Reduce 1D`
  - Confirm that the reduction completes without any issues

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
